util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+{
+	return (ct->opts.options & flags) == flags;
+}
+
+/*******************************************************************************************/
+/*                                     Data verification                                   */
+/*******************************************************************************************/
+
+void pp_fill_buf(void *buf, int size)
+{
+	char *msg_buf;
+	int msg_index;
+	static unsigned int iter = 0;
+	int i;
+
+	msg_index = ((iter++)*INTEG_SEED) % integ_alphabet_length;
+	msg_buf = (char *)buf;
+	for (i = 0; i < size; i++) {
+		msg_buf[i] = integ_alphabet[msg_index++];
+		if (msg_index >= integ_alphabet_length)
+			msg_index = 0;
+	}
+}
+
+int pp_check_buf(void *buf, int size)
+{
+	char *recv_data;
+	char c;
+	static unsigned int iter = 0;
+	int msg_index;
+	int i;
+
+	PP_DEBUG("Verifying buffer content\n");
+
+	msg_index = ((iter++)*INTEG_SEED) % integ_alphabet_length;
+	recv_data = (char *)buf;
+
+	for (i = 0; i < size; i++) {
+		c = integ_alphabet[msg_index++];
+		if (msg_index >= integ_alphabet_length)
+			msg_index = 0;
+		if (c != recv_data[i])
+			break;
+	}
+	if (i != size) {
+		PP_DEBUG("Finished veryfing buffer : content is corrupted\n");
+		printf("Error at iteration=%d size=%d byte=%d\n",
+			iter, size, i);
+		return 1;
+	}
+
+	PP_DEBUG("Buffer verified\n");
+
+	return 0;
+}
+
+/*******************************************************************************************/
+/*                                      Error handling                                     */
+/*******************************************************************************************/
+
+
+void eq_readerr(struct fid_eq *eq)
+{
+	struct fi_eq_err_entry eq_err;
+	int rd;
+
+	rd = fi_eq_readerr(eq, &eq_err, 0);
+	if (rd != sizeof(eq_err)) {
+		PP_PRINTERR("fi_eq_readerr", rd);
+	} else {
+		PP_ERR("eq_readerr: %s", fi_eq_strerror(eq, eq_err.prov_errno,
+					eq_err.err_data, NULL, 0));
+
+	}
+}
+
+void pp_process_eq_err(ssize_t rd, struct fid_eq *eq, const char *fn) {
+	if (rd == -FI_EAVAIL) {
+		eq_readerr(eq);
+	} else {
+		PP_PRINTERR(fn, rd);
+	}
+}
+
+/*******************************************************************************************/
+/*                                    Addresses handling                                   */
+/*******************************************************************************************/
+
+static int pp_getsrcaddr(char *node, char *service, struct fi_info *hints)
+{
+	struct fi_info *info;
+	int ret = 0;
+
+	ret = fi_getinfo(PP_FIVERSION, node, service, FI_SOURCE, NULL, &info);
+	if (ret) {
+		PP_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+
+	hints->src_addrlen = info->src_addrlen;
+	hints->src_addr = calloc(1, hints->src_addrlen);
+	if (!hints->src_addr) {
+		ret = -errno;
+		PP_PRINTERR("calloc", ret);
+		goto err;
+	}
+
+	memcpy(hints->src_addr, info->src_addr, hints->src_addrlen);
+	if (!hints->src_addr) {
```

Comment:
Hmm, yeah it seems that this will always evaluate to `false`. If `info->src_addr` is NULL, then this may cause a seg fault though. Maybe an assert or a check of some sort should be thrown in?